src/JsonSchema/Constraints/UndefinedConstraint.php

```diff
@@ -144,7 +144,8 @@ protected function validateCommonProperties(&$value, $schema = null, JsonPointer
             } elseif (isset($schema->required) && !is_array($schema->required)) {
                 // Draft 3 - Required attribute - e.g. "foo": {"type": "string", "required": true}
                 if ($schema->required && $value instanceof self) {
-                    $this->addError(ConstraintError::REQUIRED_D3(), $path);
+                    $propertyName = end(array_values($path->getPropertyPaths()));
```

Comment:
Yeah, `end()` and friends are a bit annoying that way, although I can understand why it's required - sometimes you need to modify the referenced element, which can't be done without a reference. Would be nice if there was a handy way to refer to the last element without requiring this kind of shenanigans, but sadly PHP doesn't provide that functionality.

You definitely get points for creativity - using `array_values` to resolve it is certainly thinking outside the box! The intent is much less obvious than just using the intermediate variable though. The extra line is annoying, but IMO worth it for making the code obvious to whoever else reads it.